msmbuilder/decomposition/ksparsetica.py

```diff
@@ -0,0 +1,207 @@
+
+
+    See Also
+    --------
+    msmbuilder.decomposition.tICA
+    msmbuilder.decomposition.SparseTICA
+
+    References
+    ----------
+    .. [1] Yuan, X-T. and Zhang, T. "Truncated Power Method for Sparse Eigenvalue Problems."
+        Journal of Machine Learning Research. Vol. 14. 2013.
+        http://www.jmlr.org/papers/volume14/yuan13a/yuan13a.pdf
+    .. [2] McGibbon, R. T. and Pande, V.S. "Identification of simple reaction
+        coordinates from complex dynamics." arXiv:1602.08776 [cond-mat.state-mech]
+        2016. https://arxiv.org/abs/1602.08776
+    .. [3] Mackey, L. "Deflation Methods for Sparse PCA." NIPS. Vol. 21. 2008.
+        http://web.stanford.edu/~lmackey/papers/deflation-nips08.pdf
+    .. [4] Noe, F. and Clementi, C. "Kinetic distance and kinetic maps from molecular dynamics
+        simulation." arXiv:1506.06259 [physics.comp-ph] https://arxiv.org/abs/1506.06259
+    """
+
+    def __init__(self, n_components=None, lag_time=1, k=5, shrinkage=None, kinetic_mapping=True):
+        super(KSparseTICA, self).__init__(n_components, lag_time=lag_time,
+                                          kinetic_mapping=kinetic_mapping, shrinkage=shrinkage)
+        self.k = k
+
+    def _truncate(self, x, k):
+        ''' given a vector x, leave its top-k absolute-value entries alone, and set the rest to 0 '''
+        not_F = np.argsort(np.abs(x))[:-k]
+        x[not_F] = 0
+        return x
+
+    def _normalize(self, x):
+        ''' given a vector x, normalize it '''
+        return x / np.linalg.norm(x)
+
+    def _truncated_power_method(self, A, x0, k, max_iter=10000, thresh=1e-8):
+        '''
+        given a matrix A, an initial guess x0, and a maximum cardinality k,
+        find the best k-sparse approximation to its dominant eigenvector
+
+        References
+        ----------
+        [1] Yuan, X-T. and Zhang, T. "Truncated Power Method for Sparse Eigenvalue Problems."
+        Journal of Machine Learning Research. Vol. 14. 2013.
+        http://www.jmlr.org/papers/volume14/yuan13a/yuan13a.pdf
+        '''
+
+        xts = [x0]
+        for t in range(max_iter):
+            xts.append(self._normalize(self._truncate(np.dot(A, xts[-1]), k)))
+            if np.linalg.norm(xts[-1] - xts[-2]) < thresh: break
+        return xts[-1]
+
+    def _solve(self):
+        if not self._is_dirty: return
+
+        # initialize
+        A, B = self.offset_correlation_, self.covariance_
+        BinvA = np.linalg.inv(B).dot(A)
+        eig_vecs = []
+
+        # recover components
+        for i in range(self.n_components):
+            x0 = self._normalize(np.random.rand(len(A)))
+            eig_vecs.append(self._truncated_power_method(BinvA, x0, k=self.k))
+            BinvA = scdeflate(BinvA, eig_vecs[-1])
+        eig_vecs = np.vstack(eig_vecs)
+
+        # compute pseudoeigenvalues
+        vs = eig_vecs.T
+        eig_vals = np.diag((vs.T.dot(A).dot(vs)).dot(np.linalg.inv(vs.T.dot(B).dot(vs))))
```

Comment:
Did a quick check and these two ways of computing eigenvalues appear to give different answers on a couple test cases. I think this may be partially because A and B tend to have very high condition numbers? In any case, I figured it was better to compute these exactly the same way as the eigenvalues would be computed in GMRQ, so that I'd be sure that `kstica.score_ == kstica.score(X_train)`.